### PR TITLE
skeletons in datumaro format

### PR DESCRIPTION
**Pull request by Eldies**
### Summary
- Added skeletons for datumaro format

### How to test
unit tests

### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/cvat-ai/datumaro/blob/develop/CHANGELOG.md)
- [ ] I have updated the [documentation](https://github.com/cvat-ai/datumaro/tree/develop/docs) accordingly
- [x] I have added tests to cover my changes
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/cvat-ai/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2022 CVAT.ai Corporation
#
# SPDX-License-Identifier: MIT
```

## Summary by CodeRabbit

- **New Features**
  - Introduced support for skeleton annotations in the Datumaro format.
  - Enhanced the `env.detect_dataset()` function to return a list of detected formats at all recursion levels.

- **Bug Fixes**
  - Improved annotation handling by allowing `Category.from_iterable` to accept a broader range of iterable types.

- **Tests**
  - Added new tests to ensure proper saving and loading of skeleton annotations.

### CODE REVIEW

**Comment by zhiltsov-max** [comment]
Could you add a skeleton annotation example in the dataset [here](https://github.com/cvat-ai/datumaro/tree/develop/tests/assets/datumaro_dataset/diverse/annotations)? We store dataset examples in different formats there.

**Comment by Eldies** [comment]
> Could you add a skeleton annotation example in the dataset [here](https://github.com/cvat-ai/datumaro/tree/develop/tests/assets/datumaro_dataset/diverse/annotations)? We store dataset examples in different formats there.

Done

**Comment by zhiltsov-max** [comment]
These examples are parsed in the tests. I suggest that this example is merged with the existing dataset or a separate parsing test is added.

**Comment by Eldies** [comment]
> These examples are parsed in the tests. I suggest that this example is merged with the existing dataset or a separate parsing test is added.

Done

File: datumaro/plugins/datumaro_format/converter.py [hunk 1]
```diff
@@ -266,6 +271,38 @@ def _convert_cuboid_3d_object(self, obj):
...
+        ][0]
+
+        def get_label_order(label_id):
+            return label_ordering.index(
+                self.categories[AnnotationType.label.name]["labels"][label_id]["name"]
+            )
+
+        points = []
+        points_attributes = []
+        for element in sorted(obj.elements, key=lambda e: get_label_order(e.label)):
```
Comment (zhiltsov-max) [inline datumaro/plugins/datumaro_format/converter.py]: It's not guaranteed that all the skeleton points are in the skeleton. The missing points should be considered having the `absent` visibility with default position 0, 0.

Comment (Eldies) [reply datumaro/plugins/datumaro_format/converter.py]: Added processing for such a case, but now there is an assert that every element of Skeleton contains exactly one point

File: datumaro/plugins/datumaro_format/converter.py [hunk 2]
```diff
@@ -143,11 +146,13 @@ def add_item(self, item: DatasetItem):
...
             elif isinstance(ann, Cuboid3d):
                 converted_ann = self._convert_cuboid_3d_object(ann)
+            elif isinstance(ann, Skeleton):
+                converted_ann = self._convert_skeleton_object(ann)
             else:
                 raise NotImplementedError()
             annotations.append(converted_ann)
 
-    def add_categories(self, categories):
+    def add_categories(self, categories: dict[AnnotationType, Categories]):
```
Comment (zhiltsov-max) [inline datumaro/plugins/datumaro_format/converter.py]: There is `dataset.components.extractor.CategoriesInfo` for this.

**Review by zhiltsov-max — APPROVED** [review]

File: datumaro/plugins/datumaro_format/converter.py (hunk 2 above)
Comment (Eldies) [reply datumaro/plugins/datumaro_format/converter.py]: fixed

**Pull request merged**